qualtran/bloqs/arithmetic/sorting.py

```diff
@@ -11,118 +11,318 @@
+
+        return {'xs': xs, 'junk': np.array(junk)}
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        return {(Comparator(self.L), self.num_comparisons)}
+
+
+@frozen
+class BitonicMerge(Bloq):
+    r"""Merge two sorted sequences of integers.
 
     Args:
-        bitsize: Number of bits used to represent each integer.
+        L: Upper-bound (excluded) on the input integers.
+        k: Number of integers in each half
+    """
+    L: SymbolicInt
+    k: SymbolicInt
+
+    def __attrs_post_init__(self):
+        k = self.k
+        if not is_symbolic(k):
+            assert not isinstance(k, sympy.Expr)
+            assert k >= 1, "length of input lists must be positive"
+            # TODO support non-power-of-two input lengths
+            assert (k & (k - 1)) == 0, "length of input lists must be a power of 2"
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [
+                Register("xs", BoundedQUInt(self.bitsize, self.L), shape=(self.k,), side=Side.LEFT),
+                Register("ys", BoundedQUInt(self.bitsize, self.L), shape=(self.k,), side=Side.LEFT),
+                Register(
+                    "result",
+                    BoundedQUInt(self.bitsize, self.L),
+                    shape=(self.k * 2,),
+                    side=Side.RIGHT,
+                ),
+                Register("junk", QBit(), shape=(self.num_comparisons,), side=Side.RIGHT),
+            ]
+        )
+
+    @cached_property
+    def bitsize(self) -> SymbolicInt:
+        return bit_length(self.L - 1)
+
+    @cached_property
+    def num_comparisons(self) -> SymbolicInt:
+        return self.k * (bit_length(self.k - 1) + 1)
+
+    def is_symbolic(self):
+        return is_symbolic(self.L, self.k)
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', xs: 'SoquetT', ys: 'SoquetT'
+    ) -> dict[str, 'SoquetT']:
+        if self.is_symbolic():
+            raise DecomposeTypeError(f"Cannot decompose symbolic {self=}")
+
+        assert isinstance(xs, np.ndarray)
+        assert isinstance(ys, np.ndarray)
+
+        k = int(self.k)
+
+        first_round_junk = []
+        for i in range(k):
+            xs[i], ys[k - 1 - i], anc = bb.add(Comparator(self.L), a=xs[i], b=ys[k - 1 - i])
+            first_round_junk.append(anc)
+
+        result = np.concatenate([xs, ys])
+        logk = int(bit_length(k - 1))
+        assert 2**logk == k
+
+        all_junks = [first_round_junk]
+        for log_offset in reversed(range(logk)):
+            result, ancs = bb.add(
+                ParallelComparators(self.L, self.k * 2, 2**log_offset), xs=result
+            )
+            all_junks.append(ancs)
+
+        junk = np.concatenate(all_junks)
+        assert len(junk) == self.num_comparisons
+
+        return {'result': result, 'junk': np.array(junk)}
+
+
+@frozen
+class BitonicSort(Bloq):
+    r"""Sort k numbers in the range {0, ..., L-1}.
```

Comment:
Use brackets `[0, L)` or `[0, L]` instead of `{0, ..., L - 1}` for consistency. 

```suggestion
    r"""Sort k numbers in the range [0, ..., L-1].
```

Reply:
Should definitely be [0, L) for exclusive upper limit

Reply:
used bitsizes instead